core/fifo_cache.go

```diff
@@ -0,0 +1,75 @@
+// (c) 2021, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package core
+
+import (
+	"sync"
+)
+
+var (
+	_ FIFOCache[int, int] = (*BufferFIFOCache[int, int])(nil)
+	_ FIFOCache[int, int] = (*NoOpFIFOCache[int, int])(nil)
+)
+
+// FIFOCache evicts the oldest element added to it after [limit] items are
+// added.
+type FIFOCache[K comparable, V any] interface {
+	Put(K, V)
+	Get(K) (V, bool)
+}
+
+// NewFIFOCache creates a new First-In-First-Out cache of size [limit].
+//
+// If a [limit] of 0 is passed as an argument, a no-op cache is returned that
+// does nothing.
+func NewFIFOCache[K comparable, V any](limit int) FIFOCache[K, V] {
+	if limit == 0 {
+		return &NoOpFIFOCache[K, V]{}
+	}
+
+	c := &BufferFIFOCache[K, V]{
+		limit: limit,
+		m:     make(map[K]V, limit),
+	}
+	c.buffer = NewBoundedBuffer[K](limit, c.remove)
```

Comment:
```suggestion
	c.buffer = NewBoundedBuffer(limit, c.remove)
```

my linter is telling me this is an unnecessary type argument